rfd/0192-oci-join.md

```diff
@@ -0,0 +1,252 @@
+- The node will then make a `RegisterUsingToken` request to the auth server and
+sends the URL and headers of the signed request as parameters (the auth server
+will be able to recreate the rest of the request).
+
+When the auth server receives a `RegisterUsingToken` request for the Oracle join method:
+
+- The auth server extracts the key ID from the provided Authorization header. The key
+ID is a string that Oracle uses to identify the caller. For instance principals,
+the key ID is a JWT prefixed by the string `ST$` (unfortunately I could not
+find docs that back this up, I found this experimentally).
+- The auth server decodes the JWT and maps the claims `opc-tenant`,
+`opc-compartment`, and `opc-instance` to the instance's tenancy ID, compartment
+ID, and instance ID respectively.
+- The auth server verifies that the compartment ID and region in the provided
+API URL match the compartment ID and region from the claims (the region can be
+extracted from the instance ID).
+- The auth server reconstructs and performs the API request; if Oracle accepts
+it, the auth server validates the tenancy ID, compartment ID (or name, found in
+the API response), and region against the Teleport provision token. If that
+passes, the node is allowed to join the cluster.
+
+#### Limitations
+
+The Oracle provision tokens will not support nested compartments, i.e. if
+compartment `foo` has a child compartment `bar` and the provision token has
+`parent_compartments: ["foo"]`, this will not allow instances in container `bar` to
+join. This is for simplicity's sake; Teleport would need to make several
+requests to the Oracle Cloud API to walk up the compartment tree from the
+compartment the instance is in, each of which would need to be signed. This
+would require a complicated back-and-forth between the auth server and the
+joining node to get signed requests for each compartment.
+
+### Security
+
+To mitigate SSRF, Teleport will sanitize the claims in the key ID JWT provided by the instance
+to ensure that they are
+[properly formed OCIDs](https://docs.oracle.com/en-us/iaas/Content/General/Concepts/identifiers.htm),
+including verifying that the region in the instance OCID is valid.
+
+To mitigate replay attacks, Teleport will verify that the JWT is not expired
+and was not issued in the future (with a leeway of 2 minutes, the same as the
+other JWT-based join methods). The Oracle API will also verify that the Date
+header in the signed request is
+[within 5 minutes](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/usingapi.htm#clock)
+of its own clock.
+
+Teleport will not verify signature of the JWT. This is because the needed JWKs will always be behind a
+[non-public API](https://docs.oracle.com/en-us/iaas/Content/APIGateway/Tasks/apigatewayusingjwttokens.htm#identityprovider).
+Instead, Teleport will trust the response from the Oracle Cloud API to know if
+it can trust the signature.
```

Comment:
again, I'm not sure we should be parsing the JWT at all, especially if we aren't verifying it

Reply:
I think there's one caveat here, which is that Oracle may be verifying the JWT when we send the request to them, and may reject the request if the JWT can't be verified. We could figure this out experimentally (send some tampered JWTs and see what happens), but because the whole JWT is undocumented it still makes me uneasy

Reply:
Oops, it didn't occur to me while writing this that Oracle might not verify the JWT at all! Fortunately @nklaassen the `authenticateClient` method you suggested works great and doesn't require the JWT, so that's what I've switched to.